### PR TITLE
Add operator detecting for c/cpp

### PR DESCRIPTION
**Pull request by TOMWT-qwq**
- Add operator detecting for c/cpp.
- Fix `FUNCTION_DECLARATION` not matching overloading operator.

### Changes

- Defined `OPERATOR` constant and added it somewhere.
- Modified something related to `FUNCTION_DECLARATION`.

### CODE REVIEW

**Comment by joshgoebel** [comment]
When you say symbol do you mean `operator`?  If so we'd do this by first putting all the operators (including multi-character ones like `!=`) into an array then using our own `regex.either` function to build the actual regex.

**Comment by joshgoebel** [comment]
I cleaned up the PR a bit to better match our coding standards... but I think the reason this hasn't been done before if the issue with `<`... sometimes it's an operator, sometimes it's the start of a template... so that case would need to be properly handled.

Right now there are a ton of failing markup tests... you'll need to go thru them and see which are operator related (which will simply need to be updated) and which are template related.

There is a line (you can uncomment) in the test/markup test code that will just "replace" the test result files with whatever we generate - it can be helpful for updating result files but if you do this please proof all the changes by hand to guarantee they are as expected.

**Comment by TOMWT-qwq** [comment]
Nothing is matched according to the tests. Is `regex.escape(x)` working properly? I can't find its defination from the package downloaded [here](https://highlightjs.org/download)

**Comment by joshgoebel** [comment]
So sorry I forgot to push the main file that does the linking... if pull and build it should work now.

**Comment by TOMWT-qwq** [comment]
ok I'll deal with those tests

**Comment by tomwt-awa** [comment]
Those tests are strong and I've found some bugs :)

Some of the failed tests are not related to cpp. I suppose they are related to the link of `regex.escape`. Do you think it's a good idea to separate these changes into two PRs?

**Comment by tomwt-awa** [comment]
I set its relevance to 0 to make aotodetect working.

Finally I've dealt with most of them

**Comment by TOMWT-qwq** [comment]
Is `->` an operator?

File: src/languages/cpp.js [hunk 1]
```diff
@@ -94,6 +94,52 @@ export default function(hljs) {
...
+    '^',
+    '&',
+    '*',
+    '<',
+    '(?<!-)>',
+    '/',
+    '-(?!>)',
+    '+',
+    '?',
+    '(?<!:):(?!:)',
```
Comment (joshgoebel) [inline src/languages/cpp.js:131]: Do we have a test case for these?  I'm pretty sure they won't work a these are string literals... `regex.escape` is going to escape all the regex stuff... we'd need to make these raw regex and then teach `escape` to just return raw regex as-is...

We should probably need a new `markup/cpp/operators.txt` test case with very minimal examples of every operator, just stuff like:

```
a == b
b > c
d * f
```

Or i'd even settle for just the full list of operators, one per line.

If there are false positives our regex are working around we'd want those included also - to make sure the behavior is correct.

---

Sadly we can't use negative look-behind until version 12 as it would be a breaking change.  What you can often do in these cases is match the offending pattern with a mode, just not assign it a scope... this effectively "skips over" the content and prevents it from being matches by other rules.

**Comment by tomwt-awa** [comment]
I don't think it's a good idea to deal with these two cases in this PR. They are not related with operator.

- `::` and `->`(and `:` in some special case, I found one in the tests) I think they should be detected as punctuation.
- template `<>`. I think they should be detected as quotes or smth else.

Anyway let's finish this stage of my work. I'll deal with them in another two PRs.

File: src/languages/cpp.js [hunk 2]
```diff
@@ -429,7 +480,28 @@ export default function(hljs) {
       /(?!switch)/,
       /(?!while)/,
       hljs.IDENT_RE,
-      regex.lookahead(/(<[^<>]+>|)\s*\(/))
+      regex.lookahead(/(<[^<>]+>|)\s*\(/)),
```
Comment (joshgoebel) [inline src/languages/cpp.js:483]: Is this related to the operator detcetion?

File: src/languages/cpp.js [hunk 3]
```diff
@@ -448,6 +521,7 @@ export default function(hljs) {
     // `return some()`, `else if()`, `(x*sum(1, 2))`
     variants: [
       {
+        returnBegin: true,
```
Comment (joshgoebel) [inline src/languages/cpp.js:524]: We're trying to get away from these, please use a look-ahead instead to avoid capturing the content rather than `returnBegin`.

**Comment by joshgoebel** [comment]
@allejo Any thoughts on whether it might be OK to just accept that `<>` as part of templates are going to be highlighted as operators (because it's too hard to distinguish) - for the increased fidelity of getting operators highlighted in general?  I've long resisted this, but maybe in cases where it wouldn't break highlighting in general we should just embrace it?

>  I'll deal with them in another two PRs.

Let's see what allejo says.  My concern is the template stuff may be very hard or even impossible to fix - so to accept this PR (separately) we'd have to be willing to accept that risk.  So we need to decide we're open to that type of brokeness long-term or if we even want to still consider it broken.

File: src/languages/cpp.js (hunk 3 above)
Comment (TOMWT-qwq) [reply src/languages/cpp.js:524]: I see.

File: src/languages/cpp.js [hunk 4]
```diff
@@ -122,7 +171,10 @@ export default function(hljs) {
     relevance: 0
   };
 
-  const FUNCTION_TITLE = regex.optional(NAMESPACE_RE) + hljs.IDENT_RE + '\\s*\\(';
+  const FUNCTION_TITLE = regex.either(
+    regex.optional(NAMESPACE_RE) + hljs.IDENT_RE + '\\s*\\(',
+    'operator\\s*' + OPERATOR_RE + '\\s*\\(',
```
Comment (joshgoebel) [inline src/languages/cpp.js:176]: The string "operator" is not part of a function title... and if you are trying to highlight the operator itself as a `title.function` in these cases it'd be better to use a multi-matcher to do that... see the documentation for `beginMatch`... so we'd match:

`"operator" [ws] [operator]` and scope each segment differently.

File: src/languages/cpp.js [hunk 5]
```diff
@@ -429,7 +480,28 @@ export default function(hljs) {
       /(?!switch)/,
       /(?!while)/,
       hljs.IDENT_RE,
-      regex.lookahead(/(<[^<>]+>|)\s*\(/))
+      regex.lookahead(/(<[^<>]+>|)\s*\(/)),
+    returnBegin: 1,
+    contains:[
```
Comment (joshgoebel) [inline src/languages/cpp.js:485]: I see now you're rewinding then parsing again, but why did you find this necessary?

File: test/markup/cpp/function-title.expect.txt [hunk 6]
```diff
@@ -1,10 +1,10 @@
-<span class="hljs-function"><span class="hljs-type">int</span> <span class="hljs-title">main</span><span class="hljs-params">(<span class="hljs-type">int</span> argc, <span class="hljs-type">char</span>** argv)</span> </span>{
-  <span class="hljs-keyword">auto</span> a = std::<span class="hljs-built_in">make_unique</span>&lt;A&gt;();
-  B *b = <span class="hljs-keyword">new</span> <span class="hljs-built_in">B</span>();
-  <span class="hljs-type">int</span> c = c * <span class="hljs-built_in">sum</span>(<span class="hljs-number">1</span>, <span class="hljs-number">2</span>);
-  <span class="hljs-keyword">if</span> (a-&gt;<span class="hljs-built_in">check1</span>())
+<span class="hljs-function"><span class="hljs-type">int</span> <span class="hljs-title">main</span><span class="hljs-params">(<span class="hljs-type">int</span> argc, <span class="hljs-type">char</span><span class="hljs-operator">*</span><span class="hljs-operator">*</span> argv)</span> </span>{
+  <span class="hljs-keyword">auto</span> a <span class="hljs-operator">=</span> std<span class="hljs-operator">::</span><span class="hljs-built_in">make_unique</span>&lt;A&gt;();
```
Comment (TOMWT-qwq) [inline test/markup/cpp/function-title.expect.txt:2]: I was trying not to match `<>` in this case with that and now I'm changing my mind and will revert that change.

File: src/languages/cpp.js (hunk 4 above)
Comment (TOMWT-qwq) [reply src/languages/cpp.js:176]:
```cpp
struct test{
	int a,b;
	inline bool operator+(const test&kkk)const{return a+kkk.b;}
};
```

I mean to highlight "operator" because I've noticed this case. I think it's a part of function title.

Comment (joshgoebel) [reply src/languages/cpp.js:176]: Did you try the tools/developer.html tool?  This already works without any additional 
changes as I can tell.

File: src/languages/cpp.js (hunk 3 above)
Comment (joshgoebel) [reply src/languages/cpp.js:504]: With v12 I'll add a note that this is deprecated in favor of literal lookahead or our `regex.lookahead` helper.

File: src/languages/cpp.js (hunk 4 above)
Comment (TOMWT-qwq) [reply src/languages/cpp.js:176]: `operator` is detected as a keyword instead of title.

Comment (joshgoebel) [reply src/languages/cpp.js:176]: OH... I see what you're expecting now... hmmm... let me noodle on that...

Comment (joshgoebel) [reply src/languages/cpp.js:176]: I see the intent, but how is it you decided `operator` is part of the 'function name' in this case?  Is that a CPP way of thinking?  Does CPP not officially declare `operator` a keyword?

I think you can make the argument the operator (in this context) is indeed `title.function` scope since it's effectively the "name" of the function, but what leads you to suggest that "operator" is also part of the function name/title? (vs just being a keyword).

Comment (TOMWT-qwq) [reply src/languages/cpp.js:176]: What's your opinion? You mean "operator" is used to decorate the real operator?

Comment (TOMWT-qwq) [reply src/languages/cpp.js:176]: Oh I think you are right after reading smth on cppreference.

Comment (TOMWT-qwq) [reply src/languages/cpp.js:176]: So how to deal with the overloaded operator? `title` or `operator`?

**Comment by tomwt-awa** [comment]
I'm going to ask you what are we going to do next btw.

- I'm planning to add punctuation detect for `;` and `::`.
- I'm planning to add namespace `::` and template `<>` to the defination of a type and a function. I hope it can work for a structure of `a::A<b::B<c::C>>>`.

Any suggestion?

File: src/languages/cpp.js (hunk 4 above)
Comment (tomwt-awa) [reply src/languages/cpp.js]: But there is no `<span class="hljs-function"></span>` now.

Comment (tomwt-awa) [reply src/languages/cpp.js:182]: Currently nothing is working properly :(

File: src/languages/cpp.js [hunk 7]
```diff
@@ -488,10 +537,9 @@ export default function(hljs) {
         relevance: 0
       },
       {
-        begin: FUNCTION_TITLE,
-        returnBegin: true,
-        contains: [ TITLE_MODE ],
-        relevance: 0
+        match: FUNCTION_TITLE,
+        scope: 'title',
+        keywords: "operator",
```
Comment (TOMWT-qwq) [inline src/languages/cpp.js:542]: 图片
It is working now.

**Comment by TOMWT-qwq** [comment]
You know, a powerful part of hljs is `contains: ["self"]`. However, a regex is not that powerful. I don't think we can recognize everything clearly like a complier now.
